File: dispatch/ios/CLAID/Sources/CLAID/RemoteFunction/FutureHandler.swift
//
//  FutureHandler.swift
//  CLAID
//
//  Created by Patrick Langer on 23.02.2025.
//


public actor FutureHandler {
    
    // The table that stores currently open futures
    private let openFutures = FuturesTable()
    
    public init() {}
    
    
    /// Creates and registers a new `Future<T>` with a freshly generated `FutureUniqueIdentifier`.
    ///
    /// - Parameter dataTypeExample: An instance of `T` just to illustrate how we might obtain a `Mutator<T>`
    ///                              (matching your `Future<T>` constructor requirement).
    /// - Returns: The newly created `Future<T>`.
    public func registerNewFuture<T: Sendable>(dataTypeExample: T) async -> Future<T> {
        // 1. Generate a unique identifier
        let uniqueIdentifier = await FutureUniqueIdentifier.makeUniqueIdentifier()
        
        // 2. Create a new Future<T>
        let future = Future<T>(
            futuresTable: openFutures,
            uniqueIdentifier: uniqueIdentifier,
            dataTypeExample: dataTypeExample
        )
        
        // 3. Store (register) the future in the FuturesTable
        //    Make sure your `FuturesTable` has an `addFuture(_:)` or similar method.
        await openFutures.addFuture(future as AbstractFuture)
        
        // 4. Return the new future
        return future
    }
    
    
    /// Look up a stored future by its unique identifier.
    ///
    /// - Returns: An `AbstractFuture?` if found, or `nil` if no future with `identifier` was found.
    public func lookupFuture(identifier: FutureUniqueIdentifier) async -> AbstractFuture? {
        // Forward the request to openFutures
        return await openFutures.lookupFuture(identifier)
    }
}

File: dispatch/python/claid/remote_function/future_handler.py
from typing import Type, TypeVar, Generic, Dict
from remote_function.abstract_future import AbstractFuture
from remote_function.future_unique_identifier import FutureUniqueIdentifier
from remote_function.future import Future
from remote_function.futures_table import FuturesTable

T = TypeVar("T")

class FutureHandler:
    def __init__(self):
        self.open_futures = FuturesTable()

    def register_new_future(self, return_data_type: Type[T]) -> Future:
        unique_identifier = FutureUniqueIdentifier.make_unique_identifier()
        future = Future(self.open_futures, unique_identifier, return_data_type)
        self.open_futures.add_future(future)
        return future

    def lookup_future(self, identifier: str) -> AbstractFuture:
        return self.open_futures.lookup_future(identifier)
